### PR TITLE
Minor: Add examples of using TreeNode with `LogicalPlan`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of #10121 and #7013 

## Rationale for this change

As @berkaysynnada says on @peter-toth 's amazing PR ' https://github.com/apache/datafusion/pull/10543

> Our TreeNode API functionality has reached a very good level. You've done a great job. Perhaps we could add documentation with code snippets to exemplify the usage of methods and their purposes, to ease the experience for less experienced users.

## What changes are included in this PR?

This PR adds examples to `LogicalPlan` and documentation to show how to use the TreeNode API

## Are these changes tested?
Yes, new doc examples

### CODE REVIEW

File: datafusion/expr/src/logical_plan/plan.rs [hunk 1]
```diff
@@ -56,19 +56,139 @@ use crate::logical_plan::tree_node::unwrap_arc;
...
+/// // Projection(name, salary)
+/// //   Filter(salary > 1000)
+/// //     TableScan(employee)
+/// # fn main() -> Result<()> {
+/// let plan = table_scan(Some("employee"), &employee_schema(), None)?
+///  .filter(col("salary").gt(lit(1000)))?
+///  .project(vec![col("name")])?
+///  .build()?;
+///
+/// // use apply to walk the plan and collect all column references
```
Comment (jonahgao) [inline datafusion/expr/src/logical_plan/plan.rs:116]:
```suggestion
/// // use apply to walk the plan and collect all expressions
```

**Review by jonahgao — APPROVED** [review]
Looks good to me!

File: datafusion/expr/src/logical_plan/plan.rs [hunk 2]
```diff
@@ -56,19 +56,139 @@ use crate::logical_plan::tree_node::unwrap_arc;
...
+/// let plan = table_scan(Some("employee"), &employee_schema(), None)?
+///  .filter(col("salary").gt(lit(1000)))?
+///  .project(vec![col("name")])?
+///  .build()?;
+///
+/// // use transform to rewrite the plan
+/// let transformed_result = plan.transform(|node| {
+///   // when we see the filter node
+///   if let LogicalPlan::Filter(mut filter) = node {
+///     // replace predicate with salary > 2000
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:169]: nice catch

**Comment by alamb** [comment]
Thanks again for the review @jonahgao

**Pull request merged**